# Creating a .truncspec file automatically to suit target ASVs

We may want to automatically create a .truncspec file to match target subsequences. For example, given a bunch of ASVs ("target ASVs") across different regions of the 16S gene, we might want to quickly get a .truncspec file that specifies positions in two reference sequences that best match the specific regions of the ASVs. Then if we then use the .truncspec file in extract16s to extract subsequences from the GTDB 16S database, we now have the GTDB 16S database truncated to suit our target ASVs.

This document outlines the process of creating a .truncspec file this way by using the `create_truncspec.py` script.

The primary purpose is to create a custom dataset to train a micro16s model, but this process could be used for other purposes as well.


---


## 1. Find a good reference sequence for Archaea and Bacteria

We want a reference sequence in the GTDB 16S databases (i.e. ssu_all_r226.fna) that is likely to be present in many of our samples. You may want to invest some time finding the most promising reference sequences with a custom method. Alternatively, the reference sequences in the `16s_reference_regions` directory are good examples. In this document, to do this quickly, we will just search the database with common ASVs. 

### Archaeal example

For example, this command searches `ssu_all_r226.fna` for this Archaeal ASV:
```bash
grep -B 1 "CACCGGCAGCTCTAGTGGTAGCAGTTTTTATTGGGCCTAAAGCGTCCGTAGCCGGTTTAATAAGTCTCTGGTGAAATCCTGCAGCTTAACTGTGGGAATTGCTGGAGATACTATTAGACTTGAGATCGGGAGAGGTTAGAGGTACTCCCAGGGTAGAGGTGAAATTCTGTAATCCTGGGAGGACCGCCTGTTGCGAAGGCGTCTGACTGGAACGATTCTGACGGTGAGGGACGAAAGCTAGGGGCGCGAAC" ssu_all_r226.fna | grep "^>"
```

That search resulted with many sequences. We might decide to use this particular sequence as the Archaeal reference sequence:
```txt
>RS_GCF_022846155.1~NZ_AP025586.1 d__Archaea;p__Methanobacteriota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanocatella;s__Methanocatella smithii [location=255094..256569] [ssu_len=1476] [contig_len=1791153]
```

Then to get the DNA from it run:
```bash
grep -A 1 "RS_GCF_022846155.1~NZ_AP025586.1" ssu_all_r226.fna
```

Which gives us:
```txt
>RS_GCF_022846155.1~NZ_AP025586.1 d__Archaea;p__Methanobacteriota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanocatella;s__Methanocatella smithii [location=255094..256569] [ssu_len=1476] [contig_len=1791153]
ATTCTGTTTGATCCTGGCAGATGCTACTGCTATTGGGATTCGATTAAGCCATGCAAGTCGAACGAGTTTAGGCTCGTGGCGTACGGCTCAGTAACACGTGGATAACCTACCCTTAGGACTGGGATAACCCTGGGAAACTGGGGATAATACTGGATAGGCAATTATTCCTGTAATGGTTTTTTGTTTAAATGTTTTTTCGCCTAAGGATGGGTCTGCGGCCGATTAGGTAGTTGGTTAGGTAATGGCTTACCAAGCCTTTGATCGGTACGGGTTGTGAGAGCAAGAGCCCGGAGATGGAACCTGAGACAAGGTTCCAGGCCCTACGGGGTGCAGCAGGCGCGAAACCTCCGCAATGTGAGAAATCGCGACGGGGGGATCCCAAGTGCCATTCTTAACGGGATGGCTTTTCATTAGTGTAAAGAGCTTTTGGAATAAGAGCTGGGCAAGACCGGTGCCAGCCGCCGCGGTAACACCGGCAGCTCTAGTGGTAGCAGTTTTTATTGGGCCTAAAGCGTCCGTAGCCGGTTTAATAAGTCTCTGGTGAAATCCTGCAGCTTAACTGTGGGAATTGCTGGAGATACTATTAGACTTGAGATCGGGAGAGGTTAGAGGTACTCCCAGGGTAGAGGTGAAATTCTGTAATCCTGGGAGGACCGCCTGTTGCGAAGGCGTCTGACTGGAACGATTCTGACGGTGAGGGACGAAAGCTAGGGGCGCGAACCGGATTAGATACCCGGGTAGTCCTAGCTGTAAACGATGCGGACTTGGTGTTGGGGTGGCTTTGAGCTGTCCCAGTGCCGAAGGGAAGCTGTTAAGTCCGCCGCCTGGGAAGTACGGTCGCAAGACTGAAACTTAAAGGAATTGGCGGGGGAGCACCACAACGCGTGGAGCCTGCGGTTTAATTGGATTCAACGCCGGACATCTCACCAGAGGCGACAGCTGTATGATAGCCAGGTTGATGACTTTGCTTGACTAGCTGAGAGGAGGTGCATGGCCGCCGTCAGCTCGTACCGTGAGGCGTCCTGTTAAGTCAGGCAACGAGCGAGACCCACGCTCTTAGTTACCAGCGGATTCTTTTTTGGATGCCGGGCACACTAAGGGGACCGCCTATGATAAATAGGAGGAAGGAGTGGACGACGGTAGGTCCGTATGCCCCGAATCCTCTGGGCAACACGCGGGCTACAATGGCTGAGACAATGGGTTCCGACGCCGAAAGGCGGAGGTAATCCTCTAAACTTAGTCGTAGTTCGGATTGAGGACTGTAACTCGTTCTCATGAAGCTGGAATGCGTAGTAATCGCGTGTCACAATCGCGCGGTGAATACGTCCCTGCTCCTTGCACACACCGCCCGTCACGCCACCCAAAAAGGGATTGGATGAGGATGTAATGTTTTGTTATATTCGAATCTAGTTTTTTTAAGGAGGGCGAAGTCGTAACAAGGTAGCCGTAGGGGAACCTGCGGCTGGATCACCTCCT
```

### Bacterial example
Then for example, we found this bacterial reference sequence:
```bash
>GB_GCA_022727485.1~JAIIWL010000593.1 d__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli [location=2..1557] [ssu_len=1556] [contig_len=1573]
AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGAAGAGCGATGGAAGCTTGCTTCTATCAATCTTAGTGGCGAACGGGTGAGTAACGCGTAATCAACCTGCCCTTCAGAGGGGGACAACAGTTGGAAACGACTGCTAATACCGCATACGATCTGACCTCGGCATCGAGGATAGATGAAAGGTGGCCTCTATTTATAAGCTATCACTGAAGGAGGGGATTGCGTCTGATTAGCTAGTTGGAGGGGTAACGGCCCACCAAGGCGATGATCAGTAGCCGGTCTGAGAGGATGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATCTTCCGCAATGGACGAAAGTCTGACGGAGCAACGCCGCGTGAGTGATGACGGCCTTCGGGTTGTAAAGCTCTGTTAATCGGGACGAAAGGCCTTCTTGCGAATAGTTAGAAGGATTGACGGTACCGGAATAGAAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGATTGGTCAGTCTGTCTTAAAAGTTCGGGGCTTAACCCCGTGATGGGATGGAAACTGCCAATCTAGAGTATCGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACGAAAACTGACGCTGAGGCGCGAAAGCCAGGGGAGCGAACGGGATTAGATACCCCGGTAGTCCTGGCCGTAAACGATGGGTACTAGGTGTAGGAGGTATCGACCCCTTCTGTGCCGGAGTTAACGCAATAAGTACCCCGCCTGGGGAGTACGACCGCAAGGTTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGTATGTGGTTTAATTCGACGCAACGCGAAGAACCTTACCAGGTCTTGACATTGATGGACAGAACTAGAGATAGTTCCTCTTCTTCGGAAGCCAGAAAACAGGTGGTGCACGGTTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCCTATCTTATGTTGCCAGCACTTCGGGTGGGAACTCATGAGAGACTGCCGCAGACAATGCGGAGGAAGGCGGGGATGACGTCAAATCATCATGCCCCTTATGACCTGGGCTACACACGTACTACAATGGGAGTTAATAGACGGAAGCAATACCGCGAGGTGGAGCAAACCCGAGAAACACTCTCTCAGTTCGGATCGTAGGCTGCAACTCGCCTACGTGAAGTCGGAATCGCTAGTAATCGCAGGTCAGCATACTGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCACGAAAGTCGGAAGTGCCCAAAGCCGGTGGGGTAACCTTCGGGAGCCAGCCGTCTAAGGTAAAGTCGATGATTGGGGTGAAGTCGTAACAAGGTAGCCGTATCGGAAGGTGCGGCTGGATCACCTCCTTT
```


---


## 2. Format full reference sequences (arc_ref_seq.txt and bac_ref_seq.txt)
Now we have to make `arc_ref_seq.txt` and `bac_ref_seq.txt` files in the `16s_reference_regions` directory that contain the full reference sequences. Just like this:

The `arc_ref_seq.txt` file:
```txt
# RS_GCF_022846155.1~NZ_AP025586.1 d__Archaea;p__Methanobacteriota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanocatella;s__Methanocatella smithii [location=255094..256569] [ssu_len=1476] [contig_len=1791153]
# (Methanocatella smithii)
# This archaeal reference sequence is from the GTDB
>FULL [len=1476bp]
ATTCTGTTTGATCCTGGCAGATGCTACTGCTATTGGGATTCGATTAAGCCATGCAAGTCGAACGAGTTTAGGCTCGTGGCGTACGGCTCAGTAACACGTGGATAACCTACCCTTAGGACTGGGATAACCCTGGGAAACTGGGGATAATACTGGATAGGCAATTATTCCTGTAATGGTTTTTTGTTTAAATGTTTTTTCGCCTAAGGATGGGTCTGCGGCCGATTAGGTAGTTGGTTAGGTAATGGCTTACCAAGCCTTTGATCGGTACGGGTTGTGAGAGCAAGAGCCCGGAGATGGAACCTGAGACAAGGTTCCAGGCCCTACGGGGTGCAGCAGGCGCGAAACCTCCGCAATGTGAGAAATCGCGACGGGGGGATCCCAAGTGCCATTCTTAACGGGATGGCTTTTCATTAGTGTAAAGAGCTTTTGGAATAAGAGCTGGGCAAGACCGGTGCCAGCCGCCGCGGTAACACCGGCAGCTCTAGTGGTAGCAGTTTTTATTGGGCCTAAAGCGTCCGTAGCCGGTTTAATAAGTCTCTGGTGAAATCCTGCAGCTTAACTGTGGGAATTGCTGGAGATACTATTAGACTTGAGATCGGGAGAGGTTAGAGGTACTCCCAGGGTAGAGGTGAAATTCTGTAATCCTGGGAGGACCGCCTGTTGCGAAGGCGTCTGACTGGAACGATTCTGACGGTGAGGGACGAAAGCTAGGGGCGCGAACCGGATTAGATACCCGGGTAGTCCTAGCTGTAAACGATGCGGACTTGGTGTTGGGGTGGCTTTGAGCTGTCCCAGTGCCGAAGGGAAGCTGTTAAGTCCGCCGCCTGGGAAGTACGGTCGCAAGACTGAAACTTAAAGGAATTGGCGGGGGAGCACCACAACGCGTGGAGCCTGCGGTTTAATTGGATTCAACGCCGGACATCTCACCAGAGGCGACAGCTGTATGATAGCCAGGTTGATGACTTTGCTTGACTAGCTGAGAGGAGGTGCATGGCCGCCGTCAGCTCGTACCGTGAGGCGTCCTGTTAAGTCAGGCAACGAGCGAGACCCACGCTCTTAGTTACCAGCGGATTCTTTTTTGGATGCCGGGCACACTAAGGGGACCGCCTATGATAAATAGGAGGAAGGAGTGGACGACGGTAGGTCCGTATGCCCCGAATCCTCTGGGCAACACGCGGGCTACAATGGCTGAGACAATGGGTTCCGACGCCGAAAGGCGGAGGTAATCCTCTAAACTTAGTCGTAGTTCGGATTGAGGACTGTAACTCGTTCTCATGAAGCTGGAATGCGTAGTAATCGCGTGTCACAATCGCGCGGTGAATACGTCCCTGCTCCTTGCACACACCGCCCGTCACGCCACCCAAAAAGGGATTGGATGAGGATGTAATGTTTTGTTATATTCGAATCTAGTTTTTTTAAGGAGGGCGAAGTCGTAACAAGGTAGCCGTAGGGGAACCTGCGGCTGGATCACCTCCT
```

The `bac_ref_seq.txt` file:
```txt
# GB_GCA_022727485.1~JAIIWL010000593.1 d__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli [location=2..1557] [ssu_len=1556] [contig_len=1573]
# (Escherichia coli)
# This bacterial reference sequence is from the GTDB
>FULL [len=1556bp]
AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGAAGAGCGATGGAAGCTTGCTTCTATCAATCTTAGTGGCGAACGGGTGAGTAACGCGTAATCAACCTGCCCTTCAGAGGGGGACAACAGTTGGAAACGACTGCTAATACCGCATACGATCTGACCTCGGCATCGAGGATAGATGAAAGGTGGCCTCTATTTATAAGCTATCACTGAAGGAGGGGATTGCGTCTGATTAGCTAGTTGGAGGGGTAACGGCCCACCAAGGCGATGATCAGTAGCCGGTCTGAGAGGATGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATCTTCCGCAATGGACGAAAGTCTGACGGAGCAACGCCGCGTGAGTGATGACGGCCTTCGGGTTGTAAAGCTCTGTTAATCGGGACGAAAGGCCTTCTTGCGAATAGTTAGAAGGATTGACGGTACCGGAATAGAAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGATTGGTCAGTCTGTCTTAAAAGTTCGGGGCTTAACCCCGTGATGGGATGGAAACTGCCAATCTAGAGTATCGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACGAAAACTGACGCTGAGGCGCGAAAGCCAGGGGAGCGAACGGGATTAGATACCCCGGTAGTCCTGGCCGTAAACGATGGGTACTAGGTGTAGGAGGTATCGACCCCTTCTGTGCCGGAGTTAACGCAATAAGTACCCCGCCTGGGGAGTACGACCGCAAGGTTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGTATGTGGTTTAATTCGACGCAACGCGAAGAACCTTACCAGGTCTTGACATTGATGGACAGAACTAGAGATAGTTCCTCTTCTTCGGAAGCCAGAAAACAGGTGGTGCACGGTTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCCTATCTTATGTTGCCAGCACTTCGGGTGGGAACTCATGAGAGACTGCCGCAGACAATGCGGAGGAAGGCGGGGATGACGTCAAATCATCATGCCCCTTATGACCTGGGCTACACACGTACTACAATGGGAGTTAATAGACGGAAGCAATACCGCGAGGTGGAGCAAACCCGAGAAACACTCTCTCAGTTCGGATCGTAGGCTGCAACTCGCCTACGTGAAGTCGGAATCGCTAGTAATCGCAGGTCAGCATACTGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCACGAAAGTCGGAAGTGCCCAAAGCCGGTGGGGTAACCTTCGGGAGCCAGCCGTCTAAGGTAAAGTCGATGATTGGGGTGAAGTCGTAACAAGGTAGCCGTATCGGAAGGTGCGGCTGGATCACCTCCTTT
```

(lines beginning with # are comments)


---


## 3. Collect all fasta files containing target ASVs 

We need to collect all fasta files containing target ASVs that we want the micro16s model to handle. So for example, all ASVs from the HMR, SAR, CMR (truncated and untruncated), etc. This could be a very large number of sequences.

Let's put them all into a directory called `target_asvs` like this:
```bash
(base) haig@haigs-pc-pop:~/Repos/micro16s/extract16s$ ls ~/Repos/micro16s/extract16s/target_asvs/
AGP.fna           CMR-Trunc_SA.fna  CMR-Trunc_ST.fna    CMR-Untrunc_SA.fna  CMR-Untrunc_ST.fna  SAR.fna
CMR-Trunc_DA.fna  CMR-Trunc_SP.fna  CMR-Untrunc_DA.fna  CMR-Untrunc_SP.fna  HMC.fna
(base) haig@haigs-pc-pop:~/Repos/micro16s/extract16s$
```


---


## 4. Run create_truncspec.py

Now we can use the `create_truncspec.py` script. This script takes the `target_asvs` directory and the reference sequence files to create a .truncspec file.

### 4.1. How it works

1. For every target ASV, check if it is contained within the full reference sequences. For those that are, record their start and end positions (1-based, inclusive).
2. Cluster redundant regions globally per domain using a greedy intersection algorithm with configurable tolerance (`TOL_BP`). This produces a small set of representative regions that cover all hits within tolerance.
3. Pair archaeal and bacterial regions by source file and length, then write the .truncspec file with computed min/max lengths.

### 4.2. Configuration

Edit the configuration constants at the top of `Scripts/create_truncspec.py`:

```python
# Directory containing target ASV FASTA files (.fna, .fasta)
TARGET_ASV_DIR = "target_asvs"

# Paths to reference sequence files
ARC_REF_PATH = "16s_reference_regions/arc_ref_seq.txt"
BAC_REF_PATH = "16s_reference_regions/bac_ref_seq.txt"

# Output truncspec filename
TRUNCSPEC_OUT_NAME = "new.truncspec"

# Output directory for intermediate files and summaries
OUT_DIR = "create_truncspec_out"

# Tolerance in bp for clustering redundant regions
TOL_BP = 10

# Buffers applied to region length for min/max bounds
MIN_LEN_BUFFER = 50
MAX_LEN_BUFFER = 50

# Maximum edit distance (mismatches + indels) for fuzzy matching
# Set to 0 to require exact matches (requires edlib if > 0)
MAX_MISMATCHES = 5

# List of input FASTA filenames to skip (exact names, e.g., ["HMC.fna"])
SKIP_INPUT_FILES = []
```

### 4.3. Running the script

```bash
cd ~/Repos/micro16s/extract16s
python Scripts/create_truncspec.py
```

### 4.4. Outputs

The script creates:

1. **`new.truncspec`** (or whatever `TRUNCSPEC_OUT_NAME` is set to) - the main output file
2. **`create_truncspec_out/`** - directory containing:
   - `{file}_hits.tsv` - per-file hit tables (ASV matches in references)
   - `{file}_missing.tsv` - ASVs that didn't match either reference
   - `arc_clusters.tsv` - archaeal cluster mapping
   - `bac_clusters.tsv` - bacterial cluster mapping
   - `summary.txt` - full summary report with counts and final regions
   - `{file}_lengths.png` - histogram plots of ASV lengths with region lengths overlaid (requires matplotlib)

### 4.5. Example output

The `.truncspec` file will look like this:

```txt
# 16S Region Truncation Specification File (.truncspec)
# Generated by create_truncspec.py
# ...

ARC_REF_SEQ_ID: RS_GCF_022846155.1~NZ_AP025586.1
BAC_REF_SEQ_ID: RS_GCF_000194175.1~NZ_AEZI02000002.1
AGP-HMC_001: arc_start=472, arc_end=725, bac_start=531, bac_end=783, min_len=243, max_len=263
CMR-Trunc_SA_002: arc_start=309, arc_end=448, bac_start=349, bac_end=529, min_len=130, max_len=190
#SAR_003: arc_start=100, arc_end=300, bac_start=NA, bac_end=NA, min_len=191, max_len=211
```

Lines starting with `#` are comments or unpaired (arc-only or bac-only) entries.
